### PR TITLE
apt: use context manager for handling apt config

### PR DESCRIPTION
**Pull request by lucasmoura**
## Proposed Commit Message
apt: use context manager for handling apt config

Some apt functions require us to change the global apt config. Since we might be changing behavior on other parts of the code that require apt functionality, we are now using a context manager to guarantee that the original system config is kept after we call an apt function

## Checklist

 - [x] I have updated or added any unit tests accordingly
 - [ ] I have updated or added any integration tests accordingly
 - [ ] I have updated or added any documentation accordingly

## Does this PR require extra reviews?

 - [ ] Yes
 - [x] No

### CODE REVIEW

File: uaclient/apt.py [hunk 1]
```diff
@@ -214,6 +215,29 @@ def get_apt_cache_policy(
...
+
+        return self.apt_func()
+
+    def __exit__(self, type, value, traceback):
+        cfg = apt_pkg.config
+        # We need to restore the apt cache configuration after creating our
+        # cache, otherwise we may break people interacting with the
+        # library after importing our modules.
+        for key in self.current_apt_cfg.keys():
+            cfg.set(key, self.current_apt_cfg[key])
```
Comment (renanrodrigo) [inline uaclient/apt.py:238]: If I am not mistaken we can simplify this by just assigning the config back to apt_pkg.
```suggestion
        apt_pkg.config = self.current_apt_cfg
```

File: uaclient/apt.py [hunk 2]
```diff
@@ -754,8 +771,10 @@ def _get_apt_config():
 
 
 def get_apt_config_keys(base_key):
-    apt_cfg = _get_apt_config()
-    return apt_cfg.list(base_key)
+    with PreserveAptCfg(_get_apt_config) as apt_cfg:
```
Comment (renanrodrigo) [inline uaclient/apt.py:774]: I don't understand why we need the ctx change here if we want the current apt config... but, if it's needed here, wouldn't it be needed in `get_apt_config_values` too?

File: uaclient/apt.py [hunk 3]
```diff
@@ -214,6 +215,29 @@ def get_apt_cache_policy(
     )
 
 
+class PreserveAptCfg:
+    def __init__(self, apt_func):
+        self.apt_func = apt_func
+        self.current_apt_cfg = {}  # Dict[str, Any]
+
+    def __enter__(self):
+        apt_pkg.init_config()
```
[the remark this answers is not in the document]
Comment (lucasmoura) [reply uaclient/apt.py]: You are right, I have removed that part of the code

File: uaclient/apt.py [hunk 4]
```diff
@@ -214,6 +215,28 @@ def get_apt_cache_policy(
...
+
+        return self.apt_func()
+
+    def __exit__(self, type, value, traceback):
+        cfg = apt_pkg.config
+        # We need to restore the apt cache configuration after creating our
+        # cache, otherwise we may break people interacting with the
+        # library after importing our modules.
+        for key in self.current_apt_cfg.keys():
+            cfg.set(key, self.current_apt_cfg[key])
```
Comment (renanrodrigo) [inline uaclient/apt.py:237]: @lucasmoura should we maybe call `init_system` here?

**Review by renanrodrigo — APPROVED** [review]

**Pull request merged**